File: app/src/main/java/vn/loitp/app/activity/customviews/wwlmusic/WWLActivityMusic.java
package vn.loitp.app.activity.customviews.wwlmusic;

import android.os.Bundle;

import com.annotation.IsFullScreen;
import com.annotation.LogTag;
import com.core.base.BaseFontActivity;
import com.views.wwlmusic.layout.LWWLMusic;
import com.views.wwlmusic.utils.LWWLMusicUiUtil;

import vn.loitp.app.R;
import vn.loitp.app.activity.customviews.wwlmusic.fragments.WWLPlaylistFragment;
import vn.loitp.app.activity.customviews.wwlmusic.fragments.WWLWatchFragment;
import vn.loitp.app.activity.customviews.wwlmusic.interfaces.FragmentHost;
import vn.loitp.app.activity.customviews.wwlmusic.utils.WWLMusicDataset;

//https://github.com/vn-ttinc/Youtube-Watch-While-Layout

@LogTag("WWLActivityMusic")
@IsFullScreen(false)
//TODO convert kotlin
public class WWLActivityMusic extends BaseFontActivity implements LWWLMusic.Listener, FragmentHost {
    private LWWLMusic LWWLMusic;
    private float mLastAlpha;
    private WWLWatchFragment watchFragment;
    private WWLPlaylistFragment playlistFragment;

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_wwl_music;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        this.LWWLMusic = findViewById(R.id.watch_while_layout);
        this.LWWLMusic.setListener(this);

        this.watchFragment = (WWLWatchFragment) getSupportFragmentManager().findFragmentById(R.id.watch_fragment);
        this.playlistFragment = (WWLPlaylistFragment) getSupportFragmentManager().findFragmentById(R.id.playlist_fragment);
    }

    @Override
    public void WWL_onSliding(float offset) {
        float alpha;
        if (offset > 2.0f) {
            alpha = this.mLastAlpha * (3.0f - offset);
        } else {
            if (offset > 1.0f) {
                alpha = 0.25f + (0.75f * (2.0f - offset));
            } else {
                alpha = 1.0f;
            }
            if (offset >= 0.0f && offset <= 1.0f) {
                LWWLMusicUiUtil.updateStatusBarAlpha(this, 1.0f - offset);
            }
        }
        this.mLastAlpha = alpha;
    }

    @Override
    public void WWL_onClicked() {
        if (this.LWWLMusic.mState == com.views.wwlmusic.layout.LWWLMusic.STATE_MINIMIZED) {
            this.LWWLMusic.maximize(false);
        }
        if (this.LWWLMusic.mState == com.views.wwlmusic.layout.LWWLMusic.STATE_MAXIMIZED) {
            this.watchFragment.toggleControls();
        }
    }

    @Override
    public void WWL_onHided() {
        this.watchFragment.stopPlay();
    }

    @Override
    protected void onDestroy() {
        this.watchFragment.stopPlay();
        super.onDestroy();
    }

    @Override
    public void WWL_minimized() {
        this.mLastAlpha = 0.0f;
        this.watchFragment.hideControls();
    }

    @Override
    public void WWL_maximized() {
        this.mLastAlpha = 1.0f;
    }

    @Override
    public void goToDetail(WWLMusicDataset.DatasetItem item) {
        if (this.LWWLMusic.mState == com.views.wwlmusic.layout.LWWLMusic.STATE_HIDED) {
            this.LWWLMusic.mState = com.views.wwlmusic.layout.LWWLMusic.STATE_MAXIMIZED;
            this.LWWLMusic.mIsFullscreen = false;
            if (this.LWWLMusic.canAnimate()) {
                this.LWWLMusic.setAnimatePos(this.LWWLMusic.mMaxY);
            }
            this.LWWLMusic.reset();
        }
        this.LWWLMusic.maximize(false);

        this.watchFragment.startPlay(item);
        if (this.playlistFragment != null) {
            this.playlistFragment.updateItem(item);
        }
    }

    @Override
    public void onVideoCollapse() {
        LWWLMusicUiUtil.showSystemUI(this);
        this.LWWLMusic.exitFullscreenToMinimize();
        this.watchFragment.switchFullscreen(false);
        this.LWWLMusic.minimize(false);
    }

    @Override
    public void onVideoFullscreen(boolean selected) {
        if (selected) {
            LWWLMusicUiUtil.hideSystemUI(this);
            this.LWWLMusic.enterFullscreen();
        } else {
            LWWLMusicUiUtil.showSystemUI(this);
            this.LWWLMusic.exitFullscreen();
        }
        this.watchFragment.switchFullscreen(selected);
    }
}

File: app/src/main/java/vn/loitp/app/activity/customviews/wwlMusic/WWLActivityMusic.kt
package vn.loitp.app.activity.customviews.wwlMusic

import android.os.Bundle
import com.loitp.annotation.IsFullScreen
import com.loitp.annotation.LogTag
import com.loitpcore.core.base.BaseFontActivity
import com.loitpcore.views.wwlMusic.layout.LWWLMusic
import com.loitpcore.views.wwlMusic.utils.LWWLMusicUiUtil.hideSystemUI
import com.loitpcore.views.wwlMusic.utils.LWWLMusicUiUtil.showSystemUI
import com.loitpcore.views.wwlMusic.utils.LWWLMusicUiUtil.updateStatusBarAlpha
import kotlinx.android.synthetic.main.activity_wwl_music.*
import vn.loitp.app.R
import vn.loitp.app.activity.customviews.wwlMusic.fragments.WWLPlaylistFragment
import vn.loitp.app.activity.customviews.wwlMusic.fragments.WWLWatchFragment
import vn.loitp.app.activity.customviews.wwlMusic.interfaces.FragmentHost
import vn.loitp.app.activity.customviews.wwlMusic.utils.WWLMusicDataset

// https://github.com/vn-ttinc/Youtube-Watch-While-Layout
@LogTag("WWLActivityMusic")
@IsFullScreen(false)
class WWLActivityMusic : BaseFontActivity(), LWWLMusic.Listener, FragmentHost {

    private var mLastAlpha = 0f
    private var frmWatch: WWLWatchFragment? = null
    private var frmPlaylist: WWLPlaylistFragment? = null

    override fun setLayoutResourceId(): Int {
        return R.layout.activity_wwl_music
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setupViews()
    }

    private fun setupViews() {
        watchWhileLayout.setListener(this)
        frmWatch = supportFragmentManager.findFragmentById(R.id.frmWatch) as WWLWatchFragment?
        frmPlaylist =
            supportFragmentManager.findFragmentById(R.id.frmPlaylist) as WWLPlaylistFragment?
    }

    override fun WWL_onSliding(offset: Float) {
        val alpha: Float
        if (offset > 2.0f) {
            alpha = mLastAlpha * (3.0f - offset)
        } else {
            alpha = if (offset > 1.0f) {
                0.25f + 0.75f * (2.0f - offset)
            } else {
                1.0f
            }
            if (offset in 0.0f..1.0f) {
                updateStatusBarAlpha(activity = this, alpha = 1.0f - offset)
            }
        }
        mLastAlpha = alpha
    }

    override fun WWL_onClicked() {
        if (watchWhileLayout.mState == LWWLMusic.STATE_MINIMIZED) {
            watchWhileLayout.maximize(false)
        }
        if (watchWhileLayout.mState == LWWLMusic.STATE_MAXIMIZED) {
            frmWatch?.toggleControls()
        }
    }

    override fun WWL_onHided() {
        frmWatch?.stopPlay()
    }

    override fun onDestroy() {
        frmWatch?.stopPlay()
        super.onDestroy()
    }

    override fun WWL_minimized() {
        mLastAlpha = 0.0f
        frmWatch?.hideControls()
    }

    override fun WWL_maximized() {
        mLastAlpha = 1.0f
    }

    override fun goToDetail(item: WWLMusicDataset.DatasetItem) {
        if (watchWhileLayout.mState == LWWLMusic.STATE_HIDED) {
            watchWhileLayout.mState = LWWLMusic.STATE_MAXIMIZED
            watchWhileLayout.mIsFullscreen = false

            watchWhileLayout.apply {
                if (this.canAnimate()) {
                    this.setAnimatePos(this.mMaxY)
                }
                this.reset()
            }
        }
        watchWhileLayout.maximize(false)
        frmWatch?.startPlay(item)
        frmPlaylist?.updateItem(item)
    }

    override fun onVideoCollapse() {
        showSystemUI(this)
        watchWhileLayout.exitFullscreenToMinimize()
        frmWatch?.switchFullscreen(false)
        watchWhileLayout.minimize(false)
    }

    override fun onVideoFullscreen(selected: Boolean) {
        if (selected) {
            hideSystemUI(this)
            watchWhileLayout.enterFullscreen()
        } else {
            showSystemUI(this)
            watchWhileLayout.exitFullscreen()
        }
        frmWatch?.switchFullscreen(selected)
    }
}
